File: test/reference-output/importExample.kt
package org.gotynoOutput

import com.fasterxml.jackson.annotation.*
import com.fasterxml.jackson.module.kotlin.*
import com.fasterxml.jackson.databind.annotation.*
import com.fasterxml.jackson.databind.*
import com.fasterxml.jackson.core.*
import com.fasterxml.jackson.databind.deser.std.*
import java.text.ParseException
import java.math.BigInteger
import kotlinx.serialization.Serializable
import org.gotynoDeclarations.BigIntegerSerializer

import org.gotynoOutput.Basic

class ImportExample {
@Serializable
data class UsesImport(
    @get:JsonProperty("recruiter")
    val recruiter: Basic.Recruiter,
    @get:JsonProperty("type")
    val type: String = "UsesImport"
) : java.io.Serializable {
    companion object {
        fun create(recruiter: Basic.Recruiter, type: String = "UsesImport"): UsesImport {
            return UsesImport(type = type, recruiter = recruiter)
        }
    }
}

@Serializable
data class HoldsSomething<T>(
    @get:JsonProperty("holdingField")
    val holdingField: T
) : java.io.Serializable {
    companion object {
        fun <T> create(holdingField: T): HoldsSomething<T> {
            return HoldsSomething(holdingField = holdingField)
        }
    }
}

@Serializable
data class StructureUsingImport(
    @get:JsonProperty("event")
    val event: Basic.Event
) : java.io.Serializable {
    companion object {
        fun create(event: Basic.Event): StructureUsingImport {
            return StructureUsingImport(event = event)
        }
    }
}

@Serializable
@JsonTypeInfo(
    use = JsonTypeInfo.Id.NAME,
    include = JsonTypeInfo.As.EXISTING_PROPERTY,
    property = "type"
)
sealed class UnionUsingImport : java.io.Serializable {
    @Serializable
    @JsonTypeName("CoolEvent")
    data class CoolEvent(val data: Basic.Event) : UnionUsingImport(), java.io.Serializable {
        val type = "CoolEvent"
    }

    @Serializable
    @JsonTypeName("Other")
    data class Other(val data: Basic.Person) : UnionUsingImport(), java.io.Serializable {
        val type = "Other"
    }
}

@Serializable
data class AllConcrete(
    @get:JsonProperty("field")
    val field: HoldsSomething<Basic.Either<Basic.Maybe<StructureUsingImport>, UnionUsingImport>>
) : java.io.Serializable {
    companion object {
        fun create(field: HoldsSomething<Basic.Either<Basic.Maybe<StructureUsingImport>, UnionUsingImport>>): AllConcrete {
            return AllConcrete(field = field)
        }
    }
}
}

File: test/reference-output/importExample.hs
{-# LANGUAGE StrictData #-}
{-# LANGUAGE TemplateHaskell #-}

module GotynoOutput.ImportExample where

import Data.Aeson (FromJSON (..), ToJSON (..))
import qualified Data.Aeson as JSON
import GHC.Generics (Generic)
import qualified Gotyno.Helpers as Helpers
import Qtility

import qualified GotynoOutput.Basic as Basic

data UsesImport = UsesImport
  { _usesImportType :: Text,
    _usesImportRecruiter :: Basic.Recruiter
  }
  deriving (Eq, Show, Generic)

deriveLensAndJSON ''UsesImport

data HoldsSomething t = HoldsSomething
  { _holdsSomethingHoldingField :: t
  }
  deriving (Eq, Show, Generic)

instance (FromJSON t) => FromJSON (HoldsSomething t) where
  parseJSON =
    JSON.genericParseJSON
      JSON.defaultOptions
        {JSON.fieldLabelModifier = drop @[] (length "_HoldsSomething") >>> lowerCaseFirst}

instance (ToJSON t) => ToJSON (HoldsSomething t) where
  toJSON =
    JSON.genericToJSON
      JSON.defaultOptions
        {JSON.fieldLabelModifier = drop @[] (length "_HoldsSomething") >>> lowerCaseFirst}

makeLenses ''HoldsSomething

data StructureUsingImport = StructureUsingImport
  { _structureUsingImportEvent :: Basic.Event
  }
  deriving (Eq, Show, Generic)

deriveLensAndJSON ''StructureUsingImport

data UnionUsingImport
  = CoolEvent Basic.Event
  | Other Basic.Person
  deriving (Eq, Show, Generic)

instance ToJSON UnionUsingImport where
  toJSON = JSON.genericToJSON $ Helpers.gotynoOptions "type"

instance FromJSON UnionUsingImport where
  parseJSON = JSON.genericParseJSON $ Helpers.gotynoOptions "type"

data AllConcrete = AllConcrete
  { _allConcreteField :: (HoldsSomething (Basic.Either' (Basic.Maybe' StructureUsingImport) UnionUsingImport))
  }
  deriving (Eq, Show, Generic)

deriveLensAndJSON ''AllConcrete
